docs/graph_engine/dataset.rst

```diff
@@ -0,0 +1,201 @@
+    >>> dataset
+    Dataset(num_blocks=70, num_rows=140, schema=<class 'str'>)
+
+    >>> pipe = dataset.window(blocks_per_window=2)
+    >>> pipe
+    DatasetPipeline(num_windows=35, num_stages=1)
+
+    >>> pipe = pipe.map_batches(transform_batch)
+    >>> pipe
+    DatasetPipeline(num_windows=35, num_stages=2)
+
+    >>> batch = next(pipe.iter_torch_batches(batch_size=batch_size))
+    >>> batch
+    {'features': tensor([[...]]), 'labels': tensor([[3.],
+            [4.]])}
+
+Graph Engine Node Sampler
+=========================
+
+In this example we use the graph engine `sample_nodes` function to generate inputs to the query function.
+Since this method uses `DatasetPipeline.from_iterable <https://docs.ray.io/en/latest/data/api/dataset_pipeline.html#creating-datasetpipelines>`_
+with a generator as input, it streams the windows instead of loading them.
+
+.. code-block:: python
+
+    >>> from ray.data import DatasetPipeline
+    >>> from deepgnn.graph_engine import SamplingStrategy
+
+    >>> cl = DistributedClient([address])
+    >>> node_batch_generator = (lambda: ray.data.from_numpy(cl.sample_nodes(140, np.array([0], dtype=np.int32), SamplingStrategy.Weighted)[0]) for _ in range(10))
+    >>> pipe = DatasetPipeline.from_iterable(node_batch_generator)
+    >>> pipe
+    DatasetPipeline(num_windows=None, num_stages=1)
+
+    >>> pipe = pipe.map_batches(transform_batch)
+    >>> pipe
+    DatasetPipeline(num_windows=None, num_stages=2)
+
+    >>> batch = next(pipe.iter_torch_batches(batch_size=2))
+    >>> batch
+    {'features': tensor([[...]]), 'labels': tensor([[...],
+            [...]])}
+
+Graph Engine Edge Sampler
+=========================
+
+In this example we use the graph engine `sample_edge` function to generate edge ids as inputs to the query function.
+Since this method uses `DatasetPipeline.from_iterable <https://docs.ray.io/en/latest/data/api/dataset_pipeline.html#creating-datasetpipelines>`_
+with a generator as input, it streams the windows instead of loading them.
+
+.. code-block:: python
+
+    >>> from ray.data import DatasetPipeline
+    >>> from deepgnn.graph_engine import SamplingStrategy
+
+    >>> cl = DistributedClient([address])
+    >>> edge_batch_generator = (lambda: ray.data.from_numpy(cl.sample_edges(140, np.array([0], dtype=np.int32), SamplingStrategy.Weighted)) for _ in range(10))
+    >>> pipe = DatasetPipeline.from_iterable(edge_batch_generator)
+    >>> pipe
+    DatasetPipeline(num_windows=None, num_stages=1)
+
+    >>> def transform_batch(idx: list) -> dict:
+    ...     return {"features": g.edge_features(idx, np.array([[0, 2]]), feature_type=np.float32), "labels": g.edge_features(idx, np.array([[1, 1]]), feature_type=np.float32)}
+    >>> pipe = pipe.map_batches(transform_batch)
+    >>> pipe
+    DatasetPipeline(num_windows=None, num_stages=2)
+
+    >>> batch = next(pipe.iter_torch_batches(batch_size=2))
+    >>> batch
+    {'features': tensor([[0., 0.],
```

Comment:
1. These zeros look like they are default stubs. Can we sample something that is different?
2. Do we need to set seed to avoid random failures in a doctest?

Reply:
This one is for the edge sampler, I'd have to make a custom dataset to avoid this.